### PR TITLE
[DOCS] Example for `LogitsProcessor` class

### PR DESCRIPTION
**Pull request by shauray8**
# What does this PR do?
Added some doc string to `RepetitionPenaltyLogitsProcessor` with some examples as well.
@gante let me know if there's anything else I should add or remove from the docs.

Fixes # (issue)
#24783

## Before submitting
- [X] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).

### CODE REVIEW

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -193,12 +193,49 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
 
 class RepetitionPenaltyLogitsProcessor(LogitsProcessor):
     r"""
-    [`LogitsProcessor`] enforcing an exponential penalty on repeated sequences.
+    [`LogitsProcessor`] prevents the repetition of previous tokens through an exponential penalty.
```
Comment (gante) [inline src/transformers/generation/logits_process.py:196]:
```suggestion
    [`LogitsProcessor`] that prevents the repetition of previous tokens through an exponential penalty. 
```

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -193,12 +193,49 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
 
 class RepetitionPenaltyLogitsProcessor(LogitsProcessor):
     r"""
-    [`LogitsProcessor`] enforcing an exponential penalty on repeated sequences.
+    [`LogitsProcessor`] prevents the repetition of previous tokens through an exponential penalty. 
+    This technique invloves adjusting the scores assigned to previously generated tokens, discouraging repetition.
```
Comment (gante) [inline src/transformers/generation/logits_process.py:197]:
```suggestion
```
(all logits processors adjust the scores, so this sentence doesn't add information :) )

File: src/transformers/generation/logits_process.py [hunk 3]
```diff
@@ -193,12 +193,49 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
 
 class RepetitionPenaltyLogitsProcessor(LogitsProcessor):
     r"""
-    [`LogitsProcessor`] enforcing an exponential penalty on repeated sequences.
+    [`LogitsProcessor`] prevents the repetition of previous tokens through an exponential penalty. 
+    This technique invloves adjusting the scores assigned to previously generated tokens, discouraging repetition. 
+    This technique shares some similarities with coverage mechanisms and other aimed at reducing repetition. 
+    During the text generation process, the probability distribution for the next token is determined using a formula that incorporates 
+    token scores based on their occurrence in the generated sequence. Tokens with higher scores are less likely to be selected.
```
Comment (gante) [inline src/transformers/generation/logits_process.py:200]:
```suggestion
```
imo this doesn't add much information either :)

File: src/transformers/generation/logits_process.py [hunk 4]
```diff
@@ -193,12 +193,49 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
 
 class RepetitionPenaltyLogitsProcessor(LogitsProcessor):
     r"""
-    [`LogitsProcessor`] enforcing an exponential penalty on repeated sequences.
+    [`LogitsProcessor`] prevents the repetition of previous tokens through an exponential penalty. 
+    This technique invloves adjusting the scores assigned to previously generated tokens, discouraging repetition. 
+    This technique shares some similarities with coverage mechanisms and other aimed at reducing repetition. 
+    During the text generation process, the probability distribution for the next token is determined using a formula that incorporates 
+    token scores based on their occurrence in the generated sequence. Tokens with higher scores are less likely to be selected. 
+    The formula takes into account a fair bunch of variables which can be further seen in the [paper](https://arxiv.org/pdf/1909.05858.pdf).
```
Comment (gante) [inline src/transformers/generation/logits_process.py:201]:
```suggestion
    The formula can be seen in the original [paper](https://arxiv.org/pdf/1909.05858.pdf).
```

File: src/transformers/generation/logits_process.py [hunk 5]
```diff
@@ -193,12 +193,49 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+
+    <Tip>
+
+    It is worth noting that the effectiveness of penalized sampling relies on the model's ability to learn a reliable distribution of 
+    tokens during its training.
+
+    </Tip>
```
Comment (gante) [inline src/transformers/generation/logits_process.py:208]:
```suggestion
```
(the same applies to all logits processors -- they all rely in good baseline model performance)

File: src/transformers/generation/logits_process.py [hunk 6]
```diff
@@ -193,12 +193,49 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
 
     Args:
         repetition_penalty (`float`):
-            The parameter for repetition penalty. 1.0 means no penalty. See [this
-            paper](https://arxiv.org/pdf/1909.05858.pdf) for more details.
+            The parameter for repetition penalty. 1.0 means no penalty. According to the paper a penalty of around 1.2 
+            yields a good balance between truthful generation and lack of repetition. 
+            See [this paper](https://arxiv.org/pdf/1909.05858.pdf) for more details.
```
Comment (gante) [inline src/transformers/generation/logits_process.py:214]: Please move "According..." to before `Args:`. It is very useful information, but the argument description should be short :D

File: src/transformers/generation/logits_process.py [hunk 7]
```diff
@@ -193,12 +193,49 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+    Examples:
+
+    ```python
+    >>> from transformers import AutoTokenizer, AutoModelForCausalLM
+    
+    >>> # Initializing the model and tokenizer for it 
+    >>> model = AutoModelForCausalLM.from_pretrained("gpt2")
+    >>> # add_prefix_space - Adds an initial space to the input. This allows to treat the leading word just as any other word 
+    >>> #(Not important for this demonstration)
+    >>> tokenizer = AutoTokenizer.from_pretrained("gpt2", add_prefix_space = True)
```
Comment (gante) [inline src/transformers/generation/logits_process.py:225]:
```suggestion
    >>> tokenizer = AutoTokenizer.from_pretrained("gpt2")
```

**Review by gante — COMMENTED** [review]
@shauray8 Thank you for opening the PR -- and great example!

I've added a few comments, but they are mostly on the same line -- let's try to condense the information, to maximize the power of the docs 💪

**Comment by gante** [comment]
You'll probably need to run `make fixup` before you next commit :)

File: src/transformers/generation/logits_process.py (hunk 6 above)
Comment (shauray8) [reply src/transformers/generation/logits_process.py]: That makes sense, Thank you for pointing out all the fixes.

**Comment by shauray8** [comment]
Previously I was having some problems with `make fixup` but now it's done and reformatted, I really like how the Makefile is structured.

**Review by gante — APPROVED** [review]
Thank you for working on it 💛 @shauray8 

Tagging a core maintainer to greenlight the merge :)

**Comment by gante** [comment]
@sgugger note: the weird part of the diff seems innocuous 🤔

**Review by sgugger — APPROVED** [review]
Thanks for your PR. Can you remove the changes in the non-relevant file from this PR though?
